plugins/user-settings-backend/src/database/UserSettingsStore.ts

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @public
+ */
+export type UserSetting = {
+  bucket: string;
+  key: string;
+  value: string;
+};
+
+/**
+ * Store definition for the user settings.
+ *
+ * @public
+ */
+export interface UserSettingsStore<Transaction> {
+  transaction<T>(fn: (tx: Transaction) => Promise<T>): Promise<T>;
+
+  get(
+    tx: Transaction,
+    opts: { userEntityRef: string; bucket: string; key: string },
+  ): Promise<UserSetting>;
+
+  set(
+    tx: Transaction,
```

Comment:
It would be nice of knex.Transaction did not have to be part of this interface, and allow me to provide my own implementation.

At the moment is assumes that I want my UserSettingsStore to be in a knex supported database.

Reply:
It's a type parameter on the interface; you can set it to `undefined` or whatever and just returning null as the transaction and have your methods ignore that parameter.

But yeah then again, not sure that there's any _need_ from a consumption pattern standpoint to ever have the transactions visible rather than happening silently inside each method one by one.

Reply:
Oh yes. I see the type there now.

I think I am not understanding why this is needed. I see there is one API route that is doing two queries, set and get. But I guess that could be done internally in the implementation by making set interface return `Promise<UserSetting>`. That would allow the implementation to do both queries within a single transaction if that is even necessary.

Reply:
The pattern is mainly to keep the usage open and to reduce the implementors functions to focus on one thing (like in other plugin implementations).
That said, I will reduce the interface a little bit more (like the bucket related methods), they were planned to preload the store, but won't be necessary anymore, see https://github.com/backstage/backstage/pull/13372#issuecomment-1228250072 . We could double check after the changes if it still makes sense to keep the flexibility.